### PR TITLE
DEV-856 implement jpa for service-analytic

### PR DESCRIPTION
**Pull request by tranlyvu**

### CODE REVIEW

File: service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    private Long id;
+
+    @Column(name = "data_id", nullable = false)
+    private Long dataId;
+
+    @Column(name = "resource_id", nullable = false)
+    private Long resourceId;
+
+    @Column(name = "download_date", nullable = false)
+    private ZonedDateTime downloadDate;
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java]: MINOR Make this value-based field transient so it is not included in the serialization of this class. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS3437)

File: service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import javax.inject.Inject;
+import javax.validation.constraints.NotNull;
+
+
+
+/**
+ * @author: Tran Ly Vu
+ */
+
+public class AnalyticController {
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java]: MINOR Remove this empty class, write its code or make it an "interface". [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2094)

File: service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package sg.ncl.service.analytic.web;
+
+
+import lombok.extern.slf4j.Slf4j;
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java]: MINOR Remove this unused import 'lombok.extern.slf4j.Slf4j'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.http.MediaType;
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java]: MINOR Remove this unused import 'org.springframework.http.MediaType'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.web.bind.annotation.*;
+
+
+import javax.inject.Inject;
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java]: MINOR Remove this unused import 'javax.inject.Inject'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import javax.validation.constraints.NotNull;
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/web/AnalyticController.java]: MINOR Remove this unused import 'javax.validation.constraints.NotNull'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service-analytic/src/main/java/sg/ncl/service/analytic/AnalyticApplication.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.context.ConfigurableApplicationContext;
+import sg.ncl.common.jpa.UseJpa;
+
+
+/**
+ * @author Tran Ly Vu
+ */
+@SpringBootApplication
+@UseJpa
+public class AnalyticApplication {
```
Comment (ChrisZhong) [inline service-analytic/src/main/java/sg/ncl/service/analytic/AnalyticApplication.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1118)

File: service-analytic/src/test/java/sg/ncl/service/analytic/AnalyticServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+
+/**
+ * @author: Tran Ly Vu.
+ */
+public class AnalyticServiceTest {
```
Comment (ChrisZhong) [inline service-analytic/src/test/java/sg/ncl/service/analytic/AnalyticServiceTest.java]: BLOCKER Add some tests to this class. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2187)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 8 issues
* BLOCKER 1 blocker
* MAJOR 1 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

File: service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
+/**
+ * @author: Tran Ly Vu
+ */
+
+
+@Entity
+@Table(name = "dataset_downloads")
```
Comment (ZHANGChunwang) [inline service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java]: Change all "dataset" to "data"

File: service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    @Column(name = "hashed_user_id", nullable = false)
+    private String hashedUserId;
+
+    @Override
+    public String toString() {
+        return "DataEntity{" +
+                "id='" + id + '\'' +
+                ", data_id=" + dataId +
+                ", resource_id=" + resourceId +
+                ", date=" + downloadDate +
```
Comment (ZHANGChunwang) [inline service-analytic/src/main/java/sg/ncl/service/analytic/data/jpa/DatasetDownloadsEntity.java]: change "date" to "download_date"

File: service-analytic/src/main/java/sg/ncl/service/analytic/domain/AnalyticService.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package sg.ncl.service.analytic.domain;
+
+/**
+ * @author: Tran Ly Vu
+ */
+public interface AnalyticService {
```
Comment (ZHANGChunwang) [inline service-analytic/src/main/java/sg/ncl/service/analytic/domain/AnalyticService.java]: add one more method addDataDownloadRecord(did, rid, uid, date)

File: service-analytic/src/main/java/sg/ncl/service/analytic/logic/AnalyticServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+import sg.ncl.service.analytic.domain.AnalyticService;
+
+/**
+ * @author: Tran Ly Vu
+ */
+@Service
+@Slf4j
+public class AnalyticServiceImpl implements AnalyticService {
+
+
```
Comment (ZHANGChunwang) [inline service-analytic/src/main/java/sg/ncl/service/analytic/logic/AnalyticServiceImpl.java]: add one more method addDataDownloadRecord(did, rid, uid, date) and implement it

File: settings.gradle [hunk 13]
```diff
@@ -17,4 +17,5 @@ include 'service-version'
 include 'service-image'
 include 'service-transmission'
 include 'service-telemetry'
+include 'service-analytic'
```
Comment (ZHANGChunwang) [inline settings.gradle]: change to service-analytics

**Comment by ZHANGChunwang** [comment]
1. You will need a DataDownloadRepository in data.jpa
2. You need to create V1_8__add_table_for_data_downloads.sql

File: service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+import java.time.ZonedDateTime;
+
+/**
+ * @author: Tran Ly Vu
+ */
+@Service
+@Slf4j
+public class AnalyticsServiceImpl implements AnalyticsService {
+
+    public void addDataDownloadRecord(Long id, Long resourceId, Long Userid, ZonedDateTime date) {
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1186)
MAJOR Add the "@Override" annotation above this method signature [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1161)
MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS00117)

File: service-analytics/src/main/java/sg/ncl/service/analytics/AnalyticsApplication.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.ConfigurableApplicationContext;
+import sg.ncl.common.jpa.UseJpa;
+
+/**
+ * @author: Tran Ly  Vu
+ */
+@SpringBootApplication
+@UseJpa
+public class AnalyticsApplication {
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/AnalyticsApplication.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1118)

File: service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package sg.ncl.service.analytics.domain;
+
+import java.time.ZonedDateTime;
+
+/**
+ * @author: Tran Ly Vu
+ */
+public interface AnalyticsService {
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java]: CRITICAL Annotate the "AnalyticsService" interface with the @FunctionalInterface annotation [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1609)

File: service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+    void addDataDownloadRecord(Long id, Long resourceId, Long Dataid, ZonedDateTime downloadDate);
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS00117)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @author: Tran Ly Vu
+ */
+@Service
+@Slf4j
+public class AnalyticsServiceImpl implements AnalyticsService {
+
+    @Override
+    public void addDataDownloadRecord(Long id, Long resourceId, Long UserId, ZonedDateTime date) {
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1186)
MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS00117)

File: service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.time.ZonedDateTime;
+
+/**
+ * @author: Tran Ly Vu
+ */
+@FunctionalInterface
+public interface AnalyticsService {
+
+    void addDataDownloadRecord(Long id, Long resourceId, Long DataId, ZonedDateTime downloadDate);
```
Comment (ChrisZhong) [inline service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS00117)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 4 issues
* CRITICAL 1 critical
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: service-analytics/src/test/java/sg/ncl/service/analytics/data/jpa/logic/AnalyticsServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,8 @@
+package sg.ncl.service.analytics.data.jpa.logic;
+
+/**
+ * Created by workshop on 12/27/2016.
+ */
+public class AnalyticsServiceImplTest {
```
Comment (ZHANGChunwang) [inline service-analytics/src/test/java/sg/ncl/service/analytics/data/jpa/logic/AnalyticsServiceImplTest.java]: BLOCKER Add some tests to this class. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2187)

**Comment by ZHANGChunwang** [comment]
SonarQube analysis reported 1 issue
* BLOCKER 1 blocker

Watch the comments in this conversation to review them.

**Comment by tranlyvu** [comment]
local test done couple of days ago

File: out/production/services-in-one/db/migration/V1_8__add_table_for_data_downloads.sql [hunk 21]
```diff
@@ -0,0 +1,18 @@
+-- -----------------------------------------------------
```
Comment (ZHANGChunwang) [inline out/production/services-in-one/db/migration/V1_8__add_table_for_data_downloads.sql]: The "out" directory should not appear in the push commit.

File: service-analytics/src/main/java/sg/ncl/service/analytics/data/jpa/DataDownloadsEntity.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+ * @author: Tran Ly Vu
+ */
+
+
+@Entity
+@Table(name = "data_downloads")
+@Getter
+@Setter
+@Slf4j
+public class DataDownloadsEntity extends AbstractEntity implements DataDownloads {
```
Comment (ZHANGChunwang) [inline service-analytics/src/main/java/sg/ncl/service/analytics/data/jpa/DataDownloadsEntity.java]: Please change to DataDownloadEntity. It should be singular format.

File: service-analytics/src/main/java/sg/ncl/service/analytics/data/jpa/DataDownloadsRepository.java [hunk 23]
```diff
@@ -0,0 +1,10 @@
+package sg.ncl.service.analytics.data.jpa;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+/**
+ * @author: Tran Ly Vu
+ */
+public interface DataDownloadsRepository extends JpaRepository<DataDownloadsEntity, Long> {
```
Comment (ZHANGChunwang) [inline service-analytics/src/main/java/sg/ncl/service/analytics/data/jpa/DataDownloadsRepository.java]: Change to DataDownloadRepository

File: service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java [hunk 24]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.time.ZonedDateTime;
+
+/**
+ * @author: Tran Ly Vu
+ */
+@FunctionalInterface
+public interface AnalyticsService {
+
+    void addDataDownloadRecord(Long id, Long resourceId, Long dataId, ZonedDateTime downloadDate);
```
Comment (ZHANGChunwang) [inline service-analytics/src/main/java/sg/ncl/service/analytics/domain/AnalyticsService.java]: 'id' is auto-generated, it should not be an argument. This function should be addDataDownloadRecord(Long dataId, Long resourceId, ZonedDateTime downloadDate，String userId)

File: service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @author: Tran Ly Vu
+ */
+@Service
+@Slf4j
+public class AnalyticsServiceImpl implements AnalyticsService {
+
+    @Override
+    public void addDataDownloadRecord(Long id, Long resourceId, Long userId, ZonedDateTime date) {
```
Comment (ZHANGChunwang) [inline service-analytics/src/main/java/sg/ncl/service/analytics/logic/AnalyticsServiceImpl.java]: Change these arguments according.

File: service-analytics/src/main/java/sg/ncl/service/analytics/AnalyticsApplication.java [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
```
Comment (ZHANGChunwang) [inline service-analytics/src/main/java/sg/ncl/service/analytics/AnalyticsApplication.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1118)

**Comment by ZHANGChunwang** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by tranlyvu** [comment]
I thought this branch was meant to implement JPA , not logic yet...

**Pull request merged**